Fix to GridSpacing in auto bed level

zenmetsu opened the pull request:
xGridSpacing and yGridSpacing as integer was causing issues with the probing grid, causing the center point of the grid to be shifted

For example, with delta probing radius = 55 and auto_bed_leveling_grid_points = 4, the following points get probed:

```
-19.000, -19.000
17.000, -19.000
17.000, 17.000
-19.000, 17.000
```

Test pattern is shifted towards the negative X and Y.  Changing the test points to 5 results in a similar shift:

```
26.000, -28.000
-1.000, -28.000
-28.000, -28.000
-28.000, -1.000
-1.000, -1.000
26.000, -1.000
53.000, -1.000
26.000, 26.000
-1.000, 26.000
-28.000, 26.000
-1.000, 53.000
```

Same offset near the origin occurs.  Also the outer rows on the negative side is getting truncated (range is -28 to 53) This is due to a floor function causing truncation.  Adding 1.0  to the xProbe and yProbe calculations will fix the offset grid and the truncation.

```
0.000, -54.000
27.000, -27.000
0.000, -27.000
-27.000, -27.000
-54.000, 0.000
-27.000, 0.000
0.000, 0.000
27.000, 0.000
54.000, 0.000
27.000, 27.000
0.000, 27.000
-27.000, 27.000
0.000, 54.000
```

zenmetsu committed:
Fix to GridSpacing in auto bed level

xGridSpacing and yGridSpacing as integer was causing issues with the probing grid, causing the center point of the grid to be shifted

For example, with delta probing radius = 55 and auto_bed_leveling_grid_points = 4, the following points get probed:

```
-19.000, -19.000
17.000, -19.000
17.000, 17.000
-19.000, 17.000
```
Test pattern is shifted towards the negative X and Y.  Changing the test points to 5 results in a similar shift:
```
26.000, -28.000
-1.000, -28.000
-28.000, -28.000
-28.000, -1.000
-1.000, -1.000
26.000, -1.000
53.000, -1.000
26.000, 26.000
-1.000, 26.000
-28.000, 26.000
-1.000, 53.000
```
Same offset near the origin occurs.  Also the outer rows on the negative side is getting truncated (range is -28 to 53) This is due to a floor function causing truncation.  Adding 1.0  to the xProbe and yProbe calculations will fix the offset grid and the truncation.

```
0.000, -54.000
27.000, -27.000
0.000, -27.000
-27.000, -27.000
-54.000, 0.000
-27.000, 0.000
0.000, 0.000
27.000, 0.000
54.000, 0.000
27.000, 27.000
0.000, 27.000
-27.000, 27.000
0.000, 54.000
```

Blue-Marlin commented:
Will `adjust_delta()` find the same indices?

zenmetsu commented:
I am not certain of your ask.  Are you asking for regression purposes or to see if `adjust_delta()` has a similar issue that needs fixing?

I can enable debugging output on `adjust_delta` to find out.  I am still finding my way around the Marlin firmware, but I did identify a fault in the way that the probing points were being calculated and fixed that accordingly. I don't see any variables utilized by `adjust_delta` which would have been altered by my changes, so I do not believe that anything would have changed.

thinkyhead commented:
The first issue here is in the way that `xGridSpacing` and `yGridSpacing` are reckoned and applied. Since they are calculated as integers, for each point you accumulate the error from rounding down. We can't just round up, because this only accumulates error in the other direction.

The second issue is the fact that the probed points start at the left/front position and go up from there. So you will always get the exact left/front position, but unless the grid spacing evenly divides into the full width the last point will always fall short.

The correct solution is to change `xGridSpacing` and `yGridSpacing` to `float`. Then for each point (assuming we only want whole values) round to the nearest integer:

``` cpp
float xProbe = floor(front_probe_bed_position + xGridSpacing * xCount + 0.5f);
. . .
float yProbe = floor(front_probe_bed_position + yGridSpacing * yCount + 0.5f);
```

zenmetsu commented:
Fair enough.  Would you like me to make the suggested changes, test, and submit a new PR?

thinkyhead commented:
Yes, but… Did you know?: You can modify the existing PR. You can add a new commit and push it. And, for bonus points you can use `git rebase -i` to combine all your commits into a single commit, then use `git push -f` to update the PR.

zenmetsu commented:
I know now!  :D 

Hey, on the subject of modifying PRs, I cannot modify the language_ru.h since it was already merged and I deleted the branch on my end, but I'm looking for clarification of what the `#define MSG_ERR_REDUNDANT_TEMP` message means.  In english, it has a string of "Redundant temperature", but the russian translation currently has a 21char string which translates to "too hot".  Just looking for context so I can properly shorten that string so that the entire file is compliant with the 20char request.

zenmetsu commented:
Updated the code and tested.

```
#define AUTO_BED_LEVELING_GRID_POINTS 7
...
SENDING:G29 V4
G29 Auto Bed Leveling
Bed X: 0.000 Y: -55.000 Z: 4.565
Bed X: 37.000 Y: -37.000 Z: 4.587
Bed X: 18.000 Y: -37.000 Z: 4.578
Bed X: 0.000 Y: -37.000 Z: 4.549
Bed X: -18.000 Y: -37.000 Z: 4.533
Bed X: -37.000 Y: -37.000 Z: 4.475
Bed X: -37.000 Y: -18.000 Z: 4.497
Bed X: -18.000 Y: -18.000 Z: 4.553
Bed X: 0.000 Y: -18.000 Z: 4.614
Bed X: 18.000 Y: -18.000 Z: 4.598
Bed X: 37.000 Y: -18.000 Z: 4.655
Bed X: 55.000 Y: 0.000 Z: 4.612
Bed X: 37.000 Y: 0.000 Z: 4.616
Bed X: 18.000 Y: 0.000 Z: 4.638
Bed X: 0.000 Y: 0.000 Z: 4.681
Bed X: -18.000 Y: 0.000 Z: 4.593
Bed X: -37.000 Y: 0.000 Z: 4.526
Bed X: -55.000 Y: 0.000 Z: 4.455
Bed X: -37.000 Y: 18.000 Z: 4.525
Bed X: -18.000 Y: 18.000 Z: 4.542
Bed X: 0.000 Y: 18.000 Z: 4.623
Bed X: 18.000 Y: 18.000 Z: 4.745
Bed X: 37.000 Y: 18.000 Z: 4.728
Bed X: 37.000 Y: 37.000 Z: 4.729
Bed X: 18.000 Y: 37.000 Z: 4.701
Bed X: 0.000 Y: 37.000 Z: 4.679
Bed X: -18.000 Y: 37.000 Z: 4.566
Bed X: -37.000 Y: 37.000 Z: 4.527
Bed X: 0.000 Y: 55.000 Z: 4.574
-0.71 -0.65 -0.59 -0.53 -0.54 -0.56 -0.52
-0.68 -0.63 -0.57 -0.55 -0.52 -0.51 -0.43
-0.66 -0.60 -0.55 -0.49 -0.50 -0.45 -0.39
-0.65 -0.57 -0.51 -0.42 -0.46 -0.48 -0.49
-0.59 -0.57 -0.56 -0.48 -0.36 -0.37 -0.39
-0.61 -0.57 -0.53 -0.42 -0.40 -0.37 -0.29
-0.63 -0.57 -0.51 -0.53 -0.44 -0.36 -0.27
```

```
#define AUTO_BED_LEVELING_GRID_POINTS 8
...
G29 Auto Bed Leveling
Bed X: -24.000 Y: -39.000 Z: 4.527
Bed X: -8.000 Y: -39.000 Z: 4.578
Bed X: 8.000 Y: -39.000 Z: 4.578
Bed X: 24.000 Y: -39.000 Z: 4.595
Bed X: 39.000 Y: -24.000 Z: 4.618
Bed X: 24.000 Y: -24.000 Z: 4.597
Bed X: 8.000 Y: -24.000 Z: 4.621
Bed X: -8.000 Y: -24.000 Z: 4.598
Bed X: -24.000 Y: -24.000 Z: 4.597
Bed X: -39.000 Y: -24.000 Z: 4.528
Bed X: -39.000 Y: -8.000 Z: 4.538
Bed X: -24.000 Y: -8.000 Z: 4.568
Bed X: -8.000 Y: -8.000 Z: 4.593
Bed X: 8.000 Y: -8.000 Z: 4.615
Bed X: 24.000 Y: -8.000 Z: 4.636
Bed X: 39.000 Y: -8.000 Z: 4.650
Bed X: 39.000 Y: 8.000 Z: 4.722
Bed X: 24.000 Y: 8.000 Z: 4.707
Bed X: 8.000 Y: 8.000 Z: 4.728
Bed X: -8.000 Y: 8.000 Z: 4.705
Bed X: -24.000 Y: 8.000 Z: 4.595
Bed X: -39.000 Y: 8.000 Z: 4.542
Bed X: -39.000 Y: 24.000 Z: 4.527
Bed X: -24.000 Y: 24.000 Z: 4.551
Bed X: -8.000 Y: 24.000 Z: 4.684
Bed X: 8.000 Y: 24.000 Z: 4.706
Bed X: 24.000 Y: 24.000 Z: 4.641
Bed X: 39.000 Y: 24.000 Z: 4.752
Bed X: 24.000 Y: 39.000 Z: 4.700
Bed X: 8.000 Y: 39.000 Z: 4.593
Bed X: -8.000 Y: 39.000 Z: 4.683
Bed X: -24.000 Y: 39.000 Z: 4.565
-0.64 -0.64 -0.64 -0.54 -0.57 -0.54 -0.53 0.00
-0.61 -0.58 -0.57 -0.52 -0.52 -0.51 -0.51 0.00
-0.64 -0.57 -0.50 -0.50 -0.48 -0.50 -0.48 0.00
-0.59 -0.56 -0.53 -0.51 -0.48 -0.46 -0.45 0.00
-0.61 -0.56 -0.51 -0.39 -0.37 -0.39 -0.38 0.00
-0.60 -0.57 -0.55 -0.42 -0.39 -0.46 -0.35 0.00
-0.64 -0.65 -0.53 -0.42 -0.51 -0.40 -0.32 0.00
0.00 0.00 0.00 0.00 0.00 0.00 0.00 0.00
```

The points that were actually probed are centered correctly, but I am not certain about the 0.00 readouts at the end when I changed the grid points to 8.  Is this normal, or a sign of a problem?

zenmetsu commented:
Aaaacctually... 

Why are you suggesting a floor for yProbe and xProbe?  Those are floats, can we just allow the raw values without a floor/ceil?

If I calculate xProbe and yProbe like this...
`float xProbe = ((float)left_probe_bed_position + xGridSpacing * (float)xCount);`

I get a nice, centered, grid pattern with non-integer values.

```
G29 Auto Bed Leveling
echo:busy: processing
Bed X: 0.000 Y: -55.000 Z: 4.565
Bed X: 36.667 Y: -36.667 Z: 4.598
Bed X: 18.333 Y: -36.667 Z: 4.581
Bed X: 0.000 Y: -36.667 Z: 4.556
Bed X: -18.333 Y: -36.667 Z: 4.581
Bed X: -36.667 Y: -36.667 Z: 4.531
Bed X: -36.667 Y: -18.333 Z: 4.499
Bed X: -18.333 Y: -18.333 Z: 4.593
Bed X: 0.000 Y: -18.333 Z: 4.592
Bed X: 18.333 Y: -18.333 Z: 4.593
Bed X: 36.667 Y: -18.333 Z: 4.611
Bed X: 55.000 Y: 0.000 Z: 4.567
Bed X: 36.667 Y: 0.000 Z: 4.572
Bed X: 18.333 Y: 0.000 Z: 4.593
Bed X: 0.000 Y: 0.000 Z: 4.568
Bed X: -18.333 Y: 0.000 Z: 4.570
Bed X: -36.667 Y: 0.000 Z: 4.527
Bed X: -55.000 Y: 0.000 Z: 4.365
Bed X: -36.667 Y: 18.333 Z: 4.526
Bed X: -18.333 Y: 18.333 Z: 4.524
Bed X: 0.000 Y: 18.333 Z: 4.658
Bed X: 18.333 Y: 18.333 Z: 4.614
Bed X: 36.667 Y: 18.333 Z: 4.661
Bed X: 36.667 Y: 36.667 Z: 4.580
Bed X: 18.333 Y: 36.667 Z: 4.555
Bed X: 0.000 Y: 36.667 Z: 4.593
Bed X: -18.333 Y: 36.667 Z: 4.533
Bed X: -36.667 Y: 36.667 Z: 4.513
Bed X: 0.000 Y: 55.000 Z: 4.529
-0.61 -0.53 -0.53 -0.53 -0.53 -0.53 -0.50
-0.66 -0.57 -0.52 -0.54 -0.52 -0.50 -0.47
-0.70 -0.60 -0.51 -0.51 -0.51 -0.49 -0.47
-0.74 -0.57 -0.53 -0.53 -0.51 -0.53 -0.53
-0.57 -0.57 -0.58 -0.44 -0.49 -0.44 -0.39
-0.61 -0.59 -0.57 -0.51 -0.54 -0.52 -0.37
-0.64 -0.60 -0.56 -0.57 -0.60 -0.64 -0.55
```

zenmetsu commented:
The problem is that if grid points is set to an even number, the outer points get dropped whether or not I floor xProbe and yProbe.  

This is because with even numbers, the points along the X and Y axis will not get probed, and the offset on each side of the axis means that the outermost points will be too far out.  In my case, the probeable radius is 55mm, which is the first value that is picked for Y.  Since no point will be along X=0, any point along Y=55 will have a distance from the center exceeding 55mm.  

My fixes address all the issues so long as grid points is odd.  I can write up a fix for odd numbered points for circular beds, but it will be a little more involved that just messing with type casting...

thinkyhead commented:
> MSG_ERR_REDUNDANT_TEMP

When using two temp sensors on the same nozzle, if they get too far apart Marlin throws an error.

thinkyhead commented:
> can we just allow the raw values without a floor/ceil?

I'll quote myself:

> (assuming we only want whole values)

Do we only want whole values? That's how it has been historically. I haven't looked into the code closely enough to know whether this matters.

thinkyhead commented:
> I cannot modify the `language_ru.h` since it was already merged and I deleted the branch on my end

Create a new branch from the latest `RCBugFix` branch and continue on.

```
git fetch upstream
git checkout upstream/RCBugFix -b my_language_ru_branch
```

thinkyhead commented:
> I get a nice, centered, grid pattern with non-integer values.

Again, I merely avoided irrationals in case it matters that we use whole numbers. If your investigation shows that this doesn't matter, then I agree this is the best course.

thinkyhead commented:
> points for circular beds

There shouldn't be much difference with a circular bed, should there? The only difference is we don't probe points that fall outside the `DELTA_PROBEABLE_RADIUS`. The only places where this might fall apart is at the extreme left/right and front/back.

thinkyhead commented:
From my testing it seems to make the most sense to enforce an [odd number of probe points](https://github.com/MarlinFirmware/Marlin/pull/4651/commits/1348830f32d32ebb811cf1e1958fe7669a643cf5) with DELTA, and/or [a larger number](https://github.com/MarlinFirmware/Marlin/pull/4651/commits/1348830f32d32ebb811cf1e1958fe7669a643cf5). Otherwise most of them will be outside the probe-able radius. The best balance comes from a 5 x 5 grid.

Also, this seems to work more properly (for probing only whole number points):

``` cpp
    float yBase = front_probe_bed_position + yGridSpacing * yCount,
          yProbe = floor(yBase + (yBase < 0 ? 0 : 0.5));
    . . .
      float xBase = left_probe_bed_position + xGridSpacing * xCount,
            xProbe = floor(xBase + (xBase < 0 ? 0 : 0.5));
```

Because maths, these will _always_ be the case for Delta:

```
Points:  4  In:  0   Out:  4
Points:  9  In:  5   Out:  4  (N, S, E, W, Center)
Points: 16  In:  4   Out: 12  (NW, NW, SW, SE)
Points: 25  In: 13   Out: 12  (2xN, 2xS, 2xE, 2xW, NE, NW, SE, SW, Center)
Points: 36  In: 16   Out: 20  (4xSW, 4xSE, 4xNW, 4xNE)
Points: 49  In: 29   Out: 20  (3xN, 3xS, 3xE, 3xW, 4xSW, 4xSE, 4xNW, 4xNE, Center)
```

thinkyhead commented:
Here's the testbed code I used in my experimentation…

``` cpp
#include <iostream>
#include <math.h>
using namespace std;

#define AUTO_BED_LEVELING_GRID_POINTS 4
#define DELTA_PRINTABLE_RADIUS 90
#define DELTA_PROBEABLE_RADIUS (DELTA_PRINTABLE_RADIUS - 10)
#define LEFT_PROBE_BED_POSITION -(DELTA_PROBEABLE_RADIUS)
#define RIGHT_PROBE_BED_POSITION DELTA_PROBEABLE_RADIUS
#define FRONT_PROBE_BED_POSITION -(DELTA_PROBEABLE_RADIUS)
#define BACK_PROBE_BED_POSITION DELTA_PROBEABLE_RADIUS

void gcode_G29() {

  const int auto_bed_leveling_grid_points = AUTO_BED_LEVELING_GRID_POINTS;

  float left_probe_bed_position = LEFT_PROBE_BED_POSITION,
       right_probe_bed_position = RIGHT_PROBE_BED_POSITION,
       front_probe_bed_position = FRONT_PROBE_BED_POSITION,
        back_probe_bed_position = BACK_PROBE_BED_POSITION;

  // probe at the points of a lattice grid
  const float xGridSpacing = (right_probe_bed_position - left_probe_bed_position) / (auto_bed_leveling_grid_points - 1),
              yGridSpacing = (back_probe_bed_position - front_probe_bed_position) / (auto_bed_leveling_grid_points - 1);

  int probePointCounter = 0, out = 0;
  bool zig = (auto_bed_leveling_grid_points & 1) ? true : false;

  for (int yCount = 0; yCount < auto_bed_leveling_grid_points; yCount++) {
    float yProbe = front_probe_bed_position + yGridSpacing * yCount,
          yOr = floor(yProbe + (yProbe < 0 ? 0 : 0.5));
    int xStart, xStop, xInc;

    if (zig) {
      xStart = 0;
      xStop = auto_bed_leveling_grid_points;
      xInc = 1;
    }
    else {
      xStart = auto_bed_leveling_grid_points - 1;
      xStop = -1;
      xInc = -1;
    }

    zig = !zig;

    for (int xCount = xStart; xCount != xStop; xCount += xInc) {
      float xProbe = left_probe_bed_position + xGridSpacing * xCount,
            xOr = floor(xProbe + (xProbe < 0 ? 0 : 0.5));
      float radius = sqrt(xProbe*xProbe + yProbe*yProbe);
      bool isout = radius > DELTA_PROBEABLE_RADIUS;
      cout << (isout ? "* " : "  ");
      cout << (yProbe == 0 ? " " : yProbe < 0 ? "S" : "N");
      cout << (xProbe == 0 ? " " : xProbe < 0 ? "W" : "E") << " ";
      cout << "xProbe, yProbe: " << xProbe << ", " << yProbe
           << "  .....  " << xOr << ", " << yOr;
      if (isout) { cout << " (" << radius << ")"; out++; }
      cout << endl;
      probePointCounter++;
    } //xProbe
  } //yProbe

  cout << "Points in: " << (probePointCounter - out) << ", out: " << out << endl;
}

int main(int argc, char const *argv[]) {
  gcode_G29();
  return 0;
}
```

zenmetsu commented:
Here is the situation...

For odd numbered, it is OK...
![screenshot 2016-08-19 07 59 15](https://cloud.githubusercontent.com/assets/6358379/17814713/f811b148-65e5-11e6-8bae-16318c0f7893.png)

For even numbered, you actually end up with GRID COUNT - 2 because points are not probed along the X and Y axis, but the math to decide spacing just divides the two axes by the GRID COUNT...

![screenshot 2016-08-19 08 20 15](https://cloud.githubusercontent.com/assets/6358379/17814768/29557ba4-65e6-11e6-952d-eef977ba60ec.png)

The fix for even numbered is to use the Pythagorean theorem to calculate the grid spacing.  The outermost points flanking the axis are what you need to calculate... these were red in the prior image since they were truncated.  The ratio of the sides will be `(GRID SPACING -1) : 1`  

Once this is corrected, you get the expected number of grid points.

![screenshot 2016-08-19 08 12 46](https://cloud.githubusercontent.com/assets/6358379/17814866/975d9866-65e6-11e6-9bc2-7c54f764fabe.png)

zenmetsu commented:
So while we could just limit it... if we really want to give the user the option to define the points, we could do it this way.  Just keep in mind that assumptions are made, namely that X and Y radii are identical.  If not, we need to take a `min(x-radius,y-radius)` to properly handle this.   Better would be to calculate based on an ellipse, but that's a lot of extra work for minimal gain.

thinkyhead commented:
I like the solution, but the way that the arguments to `G29` work, there's a (L)eft probe bed position, (R)ight, (F)ront, and (B)ack, and I wanted to continue to respect those given values, in spite of the loss of 8 points. After all, users may rightfully expect these to be taken literally, and not moved inward. In that case, we might as well drop the LRFB arguments to `G29` and just calculate everything from the number of grid points configured.

> X and Y radii are identical

That's not changing.

zenmetsu commented:
Sounds good to me.  Thanks for looking into this issue.  :)

thinkyhead commented:
@zenmetsu We can certainly continue to refine this kind of stuff. It's one area where it would be nice to help the user as much as possible. When you buy a commercial 3D printer, settings are typically locked in by the vendor so it's out-of-mind, and that can be good. But for the main Marlin project the aim is to make the firmware maximally configurable.

> …might as well drop the LRFB arguments…

I was mulling this idea over, but remembered too that with "very large" beds, sometimes you might want to just print in a corner and do leveling on that region. So I guess we'll keep 'em…
